src/rust/bitbox02/src/ui/ui.rs

```diff
@@ -195,6 +201,98 @@ where
     Component {
         component,
         is_pushed: false,
+        on_drop: Box::new(|| {}),
+        _p: PhantomData,
+    }
+}
+
+pub fn menu_create(params: MenuParams<'_>) -> Component<'_> {
+    unsafe extern "C" fn c_select_word_cb(word_idx: u8, param: *mut c_void) {
+        let callback = param as *mut Box<dyn FnMut(u8)>;
+        (*callback)(word_idx);
+    }
+
+    unsafe extern "C" fn c_continue_cancel_cb(param: *mut c_void) {
+        let callback = param as *mut Box<dyn FnMut()>;
+        (*callback)();
+    }
+
+    // We want to turn &[&str] into a C char**.
+    //
+    // Step 1: create the C strings. This var has to be alive until after menu_create() finishes,
+    // otherwise the pointers we send to menu_create() will be invalid.
+    let words: Vec<[u8; 101]> = params
+        .words
+        .iter()
+        .map(|word| crate::str_to_cstr_force!(word, 100))
+        .collect();
+    // Step two: collect pointers. This var also has to be valid until menu_create() finishes, or
+    // the pointer will be invalid.
+    let c_words: Vec<*const util::c_types::c_char> =
+        words.iter().map(|word| word.as_ptr() as _).collect();
+
+    let (select_word_cb, select_word_cb_param) = match params.select_word_cb {
+        None => (None, core::ptr::null_mut()),
+        Some(cb) => (
+            Some(c_select_word_cb as _),
+            Box::into_raw(Box::new(cb)) as *mut c_void,
+        ),
+    };
+
+    let (continue_on_last_cb, continue_on_last_cb_param) = match params.continue_on_last_cb {
+        None => (None, core::ptr::null_mut()),
+        Some(cb) => (
+            Some(c_continue_cancel_cb as _),
+            Box::into_raw(Box::new(cb)) as *mut c_void,
+        ),
+    };
+
+    let (cancel_cb, cancel_cb_param) = match params.cancel_cb {
+        None => (None, core::ptr::null_mut()),
+        Some(cb) => (
+            Some(c_continue_cancel_cb as _),
+            Box::into_raw(Box::new(cb)) as *mut c_void,
+        ),
+    };
+    let title = params
+        .title
+        .map(|title| crate::str_to_cstr_force!(title, MAX_LABEL_SIZE));
+    let component = unsafe {
+        bitbox02_sys::menu_create(
+            c_words.as_ptr(),
+            select_word_cb,
+            select_word_cb_param,
+            words.len() as _,
+            title.map_or_else(
+                || core::ptr::null(),
+                |title| crate::str_to_cstr_force!(title, MAX_LABEL_SIZE).as_ptr(),
```

Comment:
why not do it all at once above in `let title = params.title.map ...`?

Reply:
It's again drop-scrope discussion time :joy: 

If you did everything above like so:

```Rust
let title = params.title.map_or_else(
        || core::ptr::null(),
        |title| crate::str_to_cstr_force!(title, MAX_LABEL_SIZE).as_ptr(),
    );
```

The pointer would be dangling, because the cstr made in the closure is dropped as soon as the closure ends.

I have verified this in two ways:

- Demo much like the previous demo, but using a closure: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=891c88e26487977aa63c0d831bc4d250

- I tested the above snippet and indeed the menu title disappeared, presumably due to the memory error.

Reply:
What confuses me is you have str_to_cstr twice:

![image](https://user-images.githubusercontent.com/25405/95068116-1c8a7900-0705-11eb-9f69-d6cf22765097.png)

shouldn't this be:

```rust
menu_create(..., title.map_or_else(|| core::ptr::null(), title.as_ptr(), ...)
```

Reply:
:heavy_check_mark: